### PR TITLE
Throw exception on empty providers property for SerivceProvicer

### LINKED ISSUE #17559: [DX] [DependencyInjection] `Cake\Core\ServiceProvider` should throw exception when `$provides` is empty
### Description

The [documentation](https://book.cakephp.org/4/en/development/dependency-injection.html#service-providers) clearly states `Additionally those services must be defined in the $provides property.`.

But working a lot with other frameworks recently, from a DX perspecive it is very counterintuitive to not be confronted with an exception on invalid usage of a `ServiceProvider`.

My suggestion would be to thow a `LogicException` in `Cake\Core\ServiceProvider::provides` whenever `$this->provides` is empty.

### CakePHP Version

4.4, 5.0

### PR DESCRIPTION
**Pull request by Deltachaos**
Fixes #17559

### CODE REVIEW

File: src/Core/ServiceProvider.php [hunk 1]
```diff
@@ -112,6 +113,10 @@ public function register(): void
      */
     public function provides(string $id): bool
     {
+        if (empty($this->provides)) {
```
Comment (dereuromark) [inline src/Core/ServiceProvider.php]:
```suggestion
        if (!$this->provides) {
```

File: src/Core/ServiceProvider.php [hunk 2]
```diff
@@ -112,6 +113,10 @@ public function register(): void
...
+            throw new LogicException('The property $provides should contain a list with service ids for this service provider');
```
Comment (dereuromark) [inline src/Core/ServiceProvider.php]:
```suggestion
            throw new LogicException('The property `$provides` should contain a list with service ids for this service provider');
```

File: src/Core/ServiceProvider.php [hunk 3]
```diff
@@ -112,6 +113,10 @@ public function register(): void
      */
     public function provides(string $id): bool
     {
+        if (!$this->provides) {
```
Comment (ADmad) [inline src/Core/ServiceProvider.php:116]: I don't think a method specifically meant to check the existence of something should be throwing an exception. This had the potential to cause problems.

Comment (dereuromark) [reply src/Core/ServiceProvider.php:116]: I do agree however that there is a logic problem with provides() not having the initial data to even make the check make sense.
Maybe this should go into 5.x next, though.

Comment (LordSimal) [reply src/Core/ServiceProvider.php:116]: Looking at https://container.thephpleague.com/4.x/service-providers/ and https://github.com/thephpleague/container/blob/4.x/src/ServiceProvider/ServiceProviderAggregate.php#L67-L76 I am also not sure if this is correct.

The reason service providers exist in the first place is to allow lazy loading services only when they are needed.
Therefore the container needs to know which services each provider can offer without actually initializing those objects.

The `services()` method is called only after the container has checked if the provider has the requested service definition inside it.

In other words: Just because one service provider doesn't provide a requested service shouldn't prevent another provider to do so.

Comment (ADmad) [reply src/Core/ServiceProvider.php:116]: Looking at the implementation again an empty $provides property would almost always indicate a user error since it would return false for every provides() call.

Comment (dereuromark) [reply src/Core/ServiceProvider.php:116]: Jep, sure sounds like it would be a silent no-op, and as such not really useful to have.
So maybe this can help users indicate that there is sth wrong and needs to be fixed up (or this check not being used as it would never be able to return true).

Comment (ADmad) [reply src/Core/ServiceProvider.php:116]: I would just move this to 5.next instead of adding t exception in a patch release.

File: src/Core/ServiceProvider.php [hunk 4]
```diff
@@ -112,6 +113,10 @@ public function register(): void
...
+            throw new LogicException('The property `$provides` should contain a list with service ids for this service provider');
```
Comment (markstory) [inline src/Core/ServiceProvider.php]:
```suggestion
            throw new LogicException('No service providers have been registered. Register a provider before using `provides()`');
```

**Comment by dereuromark** [comment]
@Deltachaos Can you rebase this into 5.next please, the next minor release upcoming?

File: src/Core/ServiceProvider.php (hunk 3 above)
Comment (markstory) [reply src/Core/ServiceProvider.php:116]: > Looking at the implementation again an empty $services property would almost always indicate a user error since it would return false for every provides() call.

Isn't this developer error though? This exception could help save people time figuring out why cake is 'broken'.

Comment (ADmad) [reply src/Core/ServiceProvider.php:116]: Yes, that's what I meant.

**Comment by Deltachaos** [comment]
@dereuromark rebased
